src/main/scala/components/memory/memory.scala

```diff
@@ -4,9 +4,84 @@ package dinocpu
 
 import chisel3._
 import chisel3.util._
-
 import chisel3.util.experimental.loadMemoryFromFile
-import firrtl.annotations.MemoryLoadFileType
+import dinocpu.MemoryOperation._
+
+/**
+  * This is the actual memory. You should never directly use this in the CPU.
+  * This module should only be instantiated in the Top file.
+  *
+  * The I/O for this module is defined in [[MemPortBusIO]].
+  */
+
+class DualPortedCombinMemory(size: Int, memfile: String) extends BaseDualPortedMemory (size, memfile) {
+  // Instruction port
+
+  wireMemPipe(io.imem)
+
+  when (io.imem.request.valid) {
+    // Put the Request into the instruction pipe and signal that instruction memory is busy
+    val request = io.imem.request.asTypeOf(new Request)
+
+    // We should only be expecting a read from instruction memory
+    assert(request.operation === Read)
+    // Check that address is pointing to a valid location in memory
+    assert (request.address < size.U)
+
+    io.imem.response.valid        := true.B
+    io.imem.response.bits.data    := memory(request.address >> 2)
+  } .otherwise {
+    io.imem.response.valid := false.B
+  }
+
+  // Data port
+
+  wireMemPipe(io.dmem)
+
+  val memAddress = Wire(UInt(32.W))
+  val memWriteData = Wire(UInt(32.W))
+
+  // Read path
+  memAddress := io.dmem.request.bits.address
+  io.dmem.response.bits.data := memory.read(memAddress >> 2)
+  io.dmem.response.valid := true.B
+
+  // Write path
+  memWriteData := io.dmem.request.bits.writedata
+  memory(memAddress >> 2) := memWriteData
+
+  when (io.dmem.request.valid) {
+    val request = io.dmem.request.asTypeOf (new Request)
+
+    // Check that non-combin write isn't being used
+    assert (request.operation =/= Write)
+    // Check that address is pointing to a valid location in memory
+    assert (request.address < size.U)
+  } .otherwise {
+    memAddress := DontCare
```

Comment:
Hmm... This is strange. I would have expected this to also introduce a combinational loop (or we still have the possibility that we're writing bogus data to memory).

Note: I tried just moving the "read path" and "write path" above into the `when (io.dmem.request.valid)` statement and it "just worked" for me. I don't see a combinational loop.